File: gcode/src/main/java/org/qw3rtrun/p3d/g/decoder/CompositeDecoder.java
package org.qw3rtrun.p3d.g.decoder;

import lombok.RequiredArgsConstructor;
import org.qw3rtrun.p3d.g.event.GEvent;

import java.util.List;
import java.util.Optional;

@RequiredArgsConstructor
public class CompositeDecoder implements GEventDecoder<GEvent> {

    private final List<GEventDecoder> encoders;

    @Override
    public Optional<GEvent> decode(String line) {
        return encoders.stream()
                .map(enc -> enc.decode(line))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst()
                .map(GEvent.class::cast);
    }

    @Override
    public boolean test(String line) {
        return encoders.stream().anyMatch(enc -> enc.test(line));
    }
}

File: gcode/src/main/kotlin/org/qw3rtrun/p3d/g/decoder/CompositeDecoder.kt
package org.qw3rtrun.p3d.g.decoder

import org.qw3rtrun.p3d.core.msg.GEvent
import java.util.Optional

class CompositeDecoder(private val encoders: List<GEventDecoder<*>>) : GEventDecoder<GEvent> {

    @Suppress("UNCHECKED_CAST")
    override fun decode(line: String): Optional<GEvent> {
        return encoders.asSequence()
            .map { (it as GEventDecoder<GEvent>).decode(line) }
            .firstOrNull { it.isPresent } ?: Optional.empty()
    }

    override fun test(s: String): Boolean {
        return encoders.any { it.test(s) }
    }
}
